File: proto/test/Test/Schema.hs
{-# LANGUAGE OverloadedStrings #-}
-- | Tier 3 in Haskell: docs/verdict-schema.json against the goldens.
--
-- A minimal validator supporting exactly the keywords the schema uses --
-- type (including a list with "null"), properties, required,
-- additionalProperties (false or a schema), items, enum, anyOf -- and
-- refusing any other keyword, so an unsupported construct fails rather than
-- passes. Asserts: every verdict golden validates; every property path the
-- schema declares is produced by at least one golden (a field nothing
-- produces is a claim the verdict does not make); verdict_version agrees.
module Test.Schema (tests) where

import Control.Monad (forM_)
import Data.Aeson (Value (..), decodeStrict)
import qualified Data.Aeson.Key as Key
import qualified Data.Aeson.KeyMap as KM
import qualified Data.ByteString as B
import Data.Foldable (toList)
import Data.List (isSuffixOf, nub, sort, (\\))
import Data.Scientific (isInteger)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Vector as V
import Rue.Proto.Golden (Artifact (..), repoRoot)
import Rue.Proto.Tenants (artifacts)
import Rue.Proto.Verdict (verdictVersion)
import System.FilePath ((</>))
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (assertBool, assertFailure, testCase, (@?=))

tests :: IO TestTree
tests = do
  root <- repoRoot
  schemaBytes <- B.readFile (root </> "docs" </> "verdict-schema.json")
  schema <- maybe (fail "docs/verdict-schema.json does not parse") pure (decodeStrict schemaBytes)
  let goldens = [(artifactPath a, v) | a <- artifacts, "verdict.json" `isSuffixOf` artifactPath a, Right bytes <- [artifactBytes a], Just v <- [decodeStrict bytes]]
  pure $
    testGroup
      "schema"
      [ testCase "the schema uses only supported keywords" $
          case unsupported schema of
            [] -> pure ()
            ks -> assertFailure ("unsupported schema keywords: " <> show ks)
      , testCase "every verdict golden validates" $
          forM_ goldens $ \(path, v) -> case validate schema v of
            [] -> pure ()
            errs -> assertFailure (path <> ":\n" <> unlines (map ("  " <>) errs))
      , testCase "every schema property path is produced by at least one golden" $ do
          let declared = sort (nub (schemaPaths [] schema))
              produced = sort (nub (concatMap (valuePaths [] . snd) goldens))
              missing = declared \\ produced
          missing @?= []
      , testCase "verdict_version agrees" $
          forM_ goldens $ \(path, v) -> case v of
            Object o | Just (Number n) <- KM.lookup "verdict_version" o -> assertBool path (n == fromIntegral verdictVersion)
            _ -> assertFailure (path <> ": no verdict_version")
      , testCase "there are verdict goldens to validate" $
          assertBool "no verdict goldens" (not (null goldens))
      ]

supportedKeywords :: [Text]
supportedKeywords = ["$schema", "$id", "title", "description", "type", "properties", "required", "additionalProperties", "items", "enum", "anyOf"]

unsupported :: Value -> [Text]
unsupported v = case v of
  Object o ->
    [k | k <- map Key.toText (KM.keys o), k `notElem` supportedKeywords]
      <> concatMap unsupported (subschemas o)
  _ -> []

subschemas :: KM.KeyMap Value -> [Value]
subschemas o =
  concat
    [ maybe [] (\p -> case p of Object ps -> toList ps; _ -> []) (KM.lookup "properties" o)
    , maybe [] (\a -> case a of Object _ -> [a]; _ -> []) (KM.lookup "additionalProperties" o)
    , maybe [] pure (KM.lookup "items" o)
    , maybe [] (\a -> case a of Array xs -> toList xs; _ -> []) (KM.lookup "anyOf" o)
    ]

-- | Errors, or none.
validate :: Value -> Value -> [String]
validate schema v = case schema of
  Object o -> concat [typeCheck o, enumCheck o, propsCheck o, itemsCheck o, anyOfCheck o]
  _ -> ["schema is not an object"]
  where
    typeCheck o = case KM.lookup "type" o of
      Nothing -> []
      Just t ->
        let allowed = case t of
              String s -> [s]
              Array xs -> [s | String s <- toList xs]
              _ -> []
         in if any (matchesType v) allowed then [] else ["type " <> show allowed <> " does not admit " <> summary v]
    enumCheck o = case KM.lookup "enum" o of
      Just (Array xs) | v `notElem` toList xs -> ["value " <> summary v <> " not in enum"]
      _ -> []
    propsCheck o = case v of
      Object obj ->
        let props = case KM.lookup "properties" o of
              Just (Object ps) -> ps
              _ -> KM.empty
            required = case KM.lookup "required" o of
              Just (Array xs) -> [Key.fromText s | String s <- toList xs]
              _ -> []
            missing = [Key.toText k | k <- required, not (KM.member k obj)]
            extra = [Key.toText k | k <- KM.keys obj, not (KM.member k props)]
            additional = KM.lookup "additionalProperties" o
            extraErrs = case additional of
              Just (Bool False) -> ["unexpected properties " <> show extra | not (null extra)]
              Just s@(Object _) -> concat [map ((T.unpack (Key.toText k) <> ": ") <>) (validate s x) | (k, x) <- KM.toList obj, not (KM.member k props)]
              _ -> []
            propErrs = concat [map ((T.unpack (Key.toText k) <> ": ") <>) (validate s x) | (k, x) <- KM.toList obj, Just s <- [KM.lookup k props]]
         in ["missing required " <> show missing | not (null missing)] <> extraErrs <> propErrs
      _ -> []
    itemsCheck o = case (KM.lookup "items" o, v) of
      (Just s, Array xs) -> concat [map (("[" <> show i <> "]: ") <>) (validate s x) | (i, x) <- zip [0 :: Int ..] (toList xs)]
      _ -> []
    anyOfCheck o = case KM.lookup "anyOf" o of
      Just (Array alts) | not (any (null . (`validate` v)) (toList alts)) -> ["no anyOf alternative admits " <> summary v]
      _ -> []

matchesType :: Value -> Text -> Bool
matchesType v t = case (t, v) of
  ("object", Object _) -> True
  ("array", Array _) -> True
  ("string", String _) -> True
  ("boolean", Bool _) -> True
  ("null", Null) -> True
  ("integer", Number n) -> isInteger n
  ("number", Number _) -> True
  _ -> False

summary :: Value -> String
summary v = case v of
  Object _ -> "an object"
  Array _ -> "an array"
  String s -> show s
  Number n -> show n
  Bool b -> show b
  Null -> "null"

-- | Every property path the schema declares, arrays as "[]", map-valued
-- objects (additionalProperties as a schema) as "{}".
schemaPaths :: [Text] -> Value -> [Text]
schemaPaths prefix v = case v of
  Object o ->
    let here = [T.intercalate "." (reverse prefix) | not (null prefix)]
        props = case KM.lookup "properties" o of
          Just (Object ps) -> concat [schemaPaths (Key.toText k : prefix) s | (k, s) <- KM.toList ps]
          _ -> []
        additional = case KM.lookup "additionalProperties" o of
          Just s@(Object _) -> schemaPaths ("{}" : prefix) s
          _ -> []
        items = case KM.lookup "items" o of
          Just s -> schemaPaths ("[]" : prefix) s
          Nothing -> []
        alts = case KM.lookup "anyOf" o of
          Just (Array xs) -> concatMap (schemaPaths prefix) (toList xs)
          _ -> []
     in here <> props <> additional <> items <> alts
  _ -> []

-- | Every path a value populates with a non-null, and for objects and arrays
-- any, value. Keys under a map-valued object are collapsed to "{}" when the
-- schema declares no properties for them; here the two map-valued fields
-- (hosts_touched, drift_policy) are known by name.
valuePaths :: [Text] -> Value -> [Text]
valuePaths prefix v =
  let here = [T.intercalate "." (reverse prefix) | not (null prefix), v /= Null]
   in here <> case v of
        Object o -> concat [valuePaths (keyOf k : prefix) x | (k, x) <- KM.toList o]
        Array xs -> concatMap (valuePaths ("[]" : prefix)) (V.toList xs)
        _ -> []
  where
    keyOf k = if mapValued then "{}" else Key.toText k
    mapValued = case prefix of
      "hosts_touched" : _ -> True
      "drift_policy" : _ -> True
      _ -> False

File: tenants/harness/tests/schema.rs
//! Tier 3: docs/verdict-schema.json against the verdicts the crates produce.
//!
//! A minimal validator supporting exactly the keywords the schema uses --
//! type (including a list with "null"), properties, required,
//! additionalProperties (false or a schema), items, enum, anyOf -- and
//! refusing any other keyword, so an unsupported construct fails rather than
//! passes. Every verdict validates; every property path the schema declares
//! is produced by at least one verdict (a field nothing produces is a claim
//! the verdict does not make); verdict_version agrees.

use std::collections::BTreeSet;
use std::fs;

use rue_core::verdict::VERDICT_VERSION;
use rue_tenants::artifacts;
use rue_tenants::golden::repo_root;
use serde_json::Value;

const SUPPORTED: &[&str] = &[
    "$schema",
    "$id",
    "title",
    "description",
    "type",
    "properties",
    "required",
    "additionalProperties",
    "items",
    "enum",
    "anyOf",
];

fn schema_and_verdicts() -> (Value, Vec<(String, Value)>) {
    let root = repo_root().unwrap();
    let schema: Value =
        serde_json::from_slice(&fs::read(root.join("docs/verdict-schema.json")).unwrap()).unwrap();
    let verdicts: Vec<(String, Value)> = artifacts(&root)
        .into_iter()
        .filter(|a| a.path.ends_with("verdict.json"))
        .map(|a| {
            (
                a.path.clone(),
                serde_json::from_slice(a.bytes.as_ref().unwrap()).unwrap(),
            )
        })
        .collect();
    (schema, verdicts)
}

fn subschemas(o: &serde_json::Map<String, Value>) -> Vec<&Value> {
    let mut v = Vec::new();
    if let Some(Value::Object(ps)) = o.get("properties") {
        v.extend(ps.values());
    }
    if let Some(a @ Value::Object(_)) = o.get("additionalProperties") {
        v.push(a);
    }
    if let Some(i) = o.get("items") {
        v.push(i);
    }
    if let Some(Value::Array(xs)) = o.get("anyOf") {
        v.extend(xs.iter());
    }
    v
}

fn unsupported(v: &Value) -> Vec<String> {
    match v {
        Value::Object(o) => {
            let mut out: Vec<String> = o
                .keys()
                .filter(|k| !SUPPORTED.contains(&k.as_str()))
                .cloned()
                .collect();
            for s in subschemas(o) {
                out.extend(unsupported(s));
            }
            out
        }
        _ => Vec::new(),
    }
}

fn matches_type(v: &Value, t: &str) -> bool {
    match (t, v) {
        ("object", Value::Object(_))
        | ("array", Value::Array(_))
        | ("string", Value::String(_))
        | ("boolean", Value::Bool(_))
        | ("null", Value::Null)
        | ("number", Value::Number(_)) => true,
        ("integer", Value::Number(n)) => n.is_i64() || n.is_u64(),
        _ => false,
    }
}

fn summary(v: &Value) -> String {
    match v {
        Value::Object(_) => "an object".into(),
        Value::Array(_) => "an array".into(),
        other => other.to_string(),
    }
}

fn validate(schema: &Value, v: &Value) -> Vec<String> {
    let Value::Object(o) = schema else {
        return vec!["schema is not an object".into()];
    };
    let mut errs = Vec::new();
    if let Some(t) = o.get("type") {
        let allowed: Vec<&str> = match t {
            Value::String(s) => vec![s.as_str()],
            Value::Array(xs) => xs.iter().filter_map(|x| x.as_str()).collect(),
            _ => Vec::new(),
        };
        if !allowed.iter().any(|t| matches_type(v, t)) {
            errs.push(format!("type {allowed:?} does not admit {}", summary(v)));
        }
    }
    if let Some(Value::Array(xs)) = o.get("enum") {
        if !xs.contains(v) {
            errs.push(format!("value {} not in enum", summary(v)));
        }
    }
    if let Value::Object(obj) = v {
        let empty = serde_json::Map::new();
        let props = match o.get("properties") {
            Some(Value::Object(ps)) => ps,
            _ => &empty,
        };
        if let Some(Value::Array(req)) = o.get("required") {
            let missing: Vec<&str> = req
                .iter()
                .filter_map(|r| r.as_str())
                .filter(|r| !obj.contains_key(*r))
                .collect();
            if !missing.is_empty() {
                errs.push(format!("missing required {missing:?}"));
            }
        }
        let extra: Vec<&String> = obj.keys().filter(|k| !props.contains_key(*k)).collect();
        match o.get("additionalProperties") {
            Some(Value::Bool(false)) if !extra.is_empty() => {
                errs.push(format!("unexpected properties {extra:?}"))
            }
            Some(s @ Value::Object(_)) => {
                for k in &extra {
                    errs.extend(
                        validate(s, &obj[*k])
                            .into_iter()
                            .map(|e| format!("{k}: {e}")),
                    );
                }
            }
            _ => {}
        }
        for (k, x) in obj {
            if let Some(s) = props.get(k) {
                errs.extend(validate(s, x).into_iter().map(|e| format!("{k}: {e}")));
            }
        }
    }
    if let (Some(s), Value::Array(xs)) = (o.get("items"), v) {
        for (i, x) in xs.iter().enumerate() {
            errs.extend(validate(s, x).into_iter().map(|e| format!("[{i}]: {e}")));
        }
    }
    if let Some(Value::Array(alts)) = o.get("anyOf") {
        if !alts.iter().any(|alt| validate(alt, v).is_empty()) {
            errs.push(format!("no anyOf alternative admits {}", summary(v)));
        }
    }
    errs
}

/// Every property path the schema declares, arrays as "[]", map-valued
/// objects (additionalProperties as a schema) as "{}".
fn schema_paths(prefix: &[String], v: &Value, out: &mut BTreeSet<String>) {
    let Value::Object(o) = v else { return };
    if !prefix.is_empty() {
        out.insert(prefix.join("."));
    }
    let extend = |p: &[String], k: &str| -> Vec<String> {
        let mut v = p.to_vec();
        v.push(k.to_string());
        v
    };
    if let Some(Value::Object(ps)) = o.get("properties") {
        for (k, s) in ps {
            schema_paths(&extend(prefix, k), s, out);
        }
    }
    if let Some(s @ Value::Object(_)) = o.get("additionalProperties") {
        schema_paths(&extend(prefix, "{}"), s, out);
    }
    if let Some(s) = o.get("items") {
        schema_paths(&extend(prefix, "[]"), s, out);
    }
    if let Some(Value::Array(xs)) = o.get("anyOf") {
        for s in xs {
            schema_paths(prefix, s, out);
        }
    }
}

/// Every path a value populates with a non-null value. The two map-valued
/// fields (hosts_touched, drift_policy) collapse their keys to "{}".
fn value_paths(prefix: &[String], v: &Value, out: &mut BTreeSet<String>) {
    if !prefix.is_empty() && !v.is_null() {
        out.insert(prefix.join("."));
    }
    let map_valued = matches!(
        prefix.last().map(String::as_str),
        Some("hosts_touched") | Some("drift_policy")
    );
    match v {
        Value::Object(o) => {
            for (k, x) in o {
                let mut p = prefix.to_vec();
                p.push(if map_valued {
                    "{}".to_string()
                } else {
                    k.clone()
                });
                value_paths(&p, x, out);
            }
        }
        Value::Array(xs) => {
            let mut p = prefix.to_vec();
            p.push("[]".to_string());
            for x in xs {
                value_paths(&p, x, out);
            }
        }
        _ => {}
    }
}

#[test]
fn the_schema_uses_only_supported_keywords() {
    let (schema, _) = schema_and_verdicts();
    assert_eq!(unsupported(&schema), Vec::<String>::new());
}

#[test]
fn every_verdict_validates() {
    let (schema, verdicts) = schema_and_verdicts();
    assert!(!verdicts.is_empty());
    for (path, v) in &verdicts {
        let errs = validate(&schema, v);
        assert!(errs.is_empty(), "{path}:\n  {}", errs.join("\n  "));
    }
}

#[test]
fn every_schema_property_path_is_produced_by_at_least_one_verdict() {
    let (schema, verdicts) = schema_and_verdicts();
    let mut declared = BTreeSet::new();
    schema_paths(&[], &schema, &mut declared);
    let mut produced = BTreeSet::new();
    for (_, v) in &verdicts {
        value_paths(&[], v, &mut produced);
    }
    let missing: Vec<&String> = declared.difference(&produced).collect();
    assert_eq!(missing, Vec::<&String>::new());
}

#[test]
fn verdict_version_agrees() {
    let (_, verdicts) = schema_and_verdicts();
    for (path, v) in &verdicts {
        assert_eq!(
            v["verdict_version"],
            serde_json::json!(VERDICT_VERSION),
            "{path}"
        );
    }
}
